apps/service_providers/llm_service/runnables.py

```diff
@@ -267,25 +272,99 @@ def invoke(
+
+        output, ai_message_resource_file_ids = self._save_response_annotations(response)
+
         if not thread_id:
             self.state.set_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID, response.thread_id)
 
+        ai_message = self.state.save_message_to_history(output, ChatMessageType.AI)
+        if ai_message_resource_file_ids:
+            self._add_resource_files_to_message_metadata(ai_message, ai_message_resource_file_ids)
+
+        return ChainOutput(output=output, prompt_tokens=0, completion_tokens=0)
+
+    @transaction.atomic()
+    def _save_response_annotations(self, response: OpenAIAssistantFinish) -> tuple[str, dict]:
+        """
+        This makes a call to OpenAI with the `run_id` and `thread_id` to get more information about the response
+        message, specifically regarding annotations.
+        - Those of type `file_citation` cannot be read, but since we already have those files, we should be fine
+        by only storing the reference to the file i.e. its external_id
+        - Those of type `file_path` are generated and can be downloaded. A file is created in OCS for each of these
+
+        """
+        from apps.assistants.sync import get_and_store_openai_file
+
+        client = self.state.get_openai_assistant().client
+        page: SyncCursorPage = client.beta.threads.messages.list(response.thread_id, run_id=response.run_id)
+        # We use a run to get a response, so it's safe to assume a run will have a single message
+        message: Message = page.data[0]
+        message_content = message.content[0]
+        annotations = message_content.text.annotations
         output = response.return_values["output"]
-        self.state.save_message_to_history(output, ChatMessageType.AI)
+        resource_file_ids = defaultdict(list)
+        generated_files = []
+        session_id = self.state.session.id
+        team_slug = self.state.session.team.slug
 
-        return ChainOutput(output=response.return_values["output"], prompt_tokens=0, completion_tokens=0)
+        assistant_file_ids = ToolResources.objects.filter(assistant=self.state.experiment.assistant).values_list(
+            "files"
+        )
+        assistant_files_ids = File.objects.filter(
+            team__slug=team_slug, id__in=models.Subquery(assistant_file_ids)
+        ).values_list("external_id", flat=True)
+
+        for idx, annotation in enumerate(annotations):
+            file_id = None
+            file_ref_text = annotation.text
+            if annotation.type == "file_citation":
+                file_citation = annotation.file_citation
+                file_id = file_citation.file_id
+
+                file_link = ""
+                if file_id not in assistant_files_ids:
+                    # Assistant level files should have an empty link
+                    team_slug = self.state.session.team.slug
+                    file = File.objects.get(external_id=file_id, team__slug=team_slug)
```

Comment:
This could error if the file wasn't uploaded via OCS

Reply:
Fair, I'll handle that case, thanks

Reply:
https://github.com/dimagi/open-chat-studio/pull/545/commits/f045b62a90311d185df691e3c55e616cf21972c5 and https://github.com/dimagi/open-chat-studio/pull/545/commits/45355729981087f14e55c34614da3b3f35a000ed